src/librustc/infer/region_inference/mod.rs

```diff
@@ -213,6 +215,7 @@ pub struct RegionVarBindings<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     // record the fact that `'a <= 'b` is implied by the fn signature,
     // and then ignore the constraint when solving equations. This is
     // a bit of a hack but seems to work.
+    early_bound_givens: RefCell<FxHashSet<(ty::EarlyBoundRegion, ty::RegionVid)>>,
```

Comment:
maybe combine these by changing the type to `RefCell<FxHashSet<(ty::Region<'tcx>, ty::RegionVid)>>`, where the region is either an early-bound or free?